### PR TITLE
Frontend ipv6 support

### PR DESCRIPTION
**Pull request by Luke-Smartnews**
#### What this PR does
Add `instance_enable_ipv6` option to query-frontend to support IPv6

#### Which issue(s) this PR fixes or relates to

Fixes [#<6111>](https://github.com/grafana/mimir/issues/6111)

#### Checklist

- [ ] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by charleskorn** [comment]
Could you please sign the CLA @Luke-Smartnews? Once that's signed, I'll review this.

**Comment by Luke-Smartnews** [comment]
@charleskorn 
resolved conflicts and  signed the CLA

File: operations/helm/charts/mimir-distributed/values.yaml [hunk 1]
```diff
@@ -2709,7 +2709,8 @@ gateway:
 
           proxy_read_timeout 300;
           server {
-            listen {{ include "mimir.serverHttpListenPort" . }};
+            listen 8080;
```
[the remark this answers is not in the document]
Comment (Luke-Smartnews) [reply operations/helm/charts/mimir-distributed/values.yaml]: done

File: pkg/frontend/config.go [hunk 2]
```diff
@@ -13,12 +13,12 @@ import (
 	"github.com/pkg/errors"
 	"github.com/prometheus/client_golang/prometheus"
 
+	"github.com/grafana/dskit/netutil"
```
Comment (krajorama) [inline pkg/frontend/config.go:16]: Hi, linter fails due to wrong order, it should go to line 13. You can run `make lint` to verify.

```
diff --git a/pkg/frontend/config.go b/pkg/frontend/config.go
index c84acec1d..e3ba77516 100644
--- a/pkg/frontend/config.go
+++ b/pkg/frontend/config.go
@@ -10,10 +10,10 @@ import (
        "net/http"
 
        "github.com/go-kit/log"
+       "github.com/grafana/dskit/netutil"
        "github.com/pkg/errors"
        "github.com/prometheus/client_golang/prometheus"
 
-       "github.com/grafana/dskit/netutil"
        "github.com/grafana/mimir/pkg/frontend/querymiddleware"
        "github.com/grafana/mimir/pkg/frontend/transport"
        v1 "github.com/grafana/mimir/pkg/frontend/v1"
```

File: pkg/frontend/v2/frontend.go [hunk 3]
```diff
@@ -45,7 +45,8 @@ type Config struct {
 	GRPCClientConfig  grpcclient.Config `yaml:"grpc_client_config" doc:"description=Configures the gRPC client used to communicate between the query-frontends and the query-schedulers."`
 
 	// Used to find local IP address, that is sent to scheduler and querier-worker.
-	InfNames []string `yaml:"instance_interface_names" category:"advanced" doc:"default=[<private network interfaces>]"`
+	InfNames   []string `yaml:"instance_interface_names" category:"advanced" doc:"default=[<private network interfaces>]"`
+	EnableIPv6 bool     `yaml:"instance_enable_ipv6" category:"advanced"`
```
Comment (krajorama) [inline pkg/frontend/v2/frontend.go:49]: Needs to be initialized in `RegisterFlags` bellow.

**Review by krajorama — COMMENTED** [review]
To let the tests and doc validation pass, you'll have to run `make reference-help` and `make doc`.

The one thing that bugs me is that all other components have the configuration parameter under the "ring" sub-section, but of course the frontend doesn't join the ring. So it's probably ok.

File: pkg/frontend/v2/frontend.go (hunk 3 above)
Comment (Luke-Smartnews) [reply pkg/frontend/v2/frontend.go:49]: Done

File: pkg/frontend/v2/frontend.go [hunk 4]
```diff
@@ -61,6 +62,7 @@ func (cfg *Config) RegisterFlags(f *flag.FlagSet, logger log.Logger) {
 	f.IntVar(&cfg.WorkerConcurrency, "query-frontend.scheduler-worker-concurrency", 5, "Number of concurrent workers forwarding queries to single query-scheduler.")
 
 	cfg.InfNames = netutil.PrivateNetworkInterfacesWithFallback([]string{"eth0", "en0"}, logger)
+	f.BoolVar(&cfg.EnableIPv6, "instance-enable-ipv6", false, "Enable using a IPv6 instance address. (default false)")
```
Comment (krajorama) [inline pkg/frontend/v2/frontend.go:65]:
```suggestion
	f.BoolVar(&cfg.EnableIPv6, "query-frontend.instance-enable-ipv6", false, "Enable using a IPv6 instance address. (default false)")
```

Comment (krajorama) [reply pkg/frontend/v2/frontend.go:65]: And please be aware that the `make reference-help` and `make doc` needs to be re-run

Comment (Luke-Smartnews) [reply pkg/frontend/v2/frontend.go]: thanks, updated doc

**Review by dimitarvdimitrov — APPROVED** [review]
LGTM

File: cmd/mimir/help-all.txt.tmpl [hunk 5]
```diff
@@ -1751,6 +1751,8 @@ Usage of ./cmd/mimir/mimir:
     	Override the expected name on the server certificate.
   -query-frontend.instance-addr string
     	IP address to advertise to the querier (via scheduler) (default is auto-detected from network interfaces).
+  -query-frontend.instance-enable-ipv6
+    	Enable using a IPv6 instance address. (default false)
```
Comment (charleskorn) [inline cmd/mimir/help-all.txt.tmpl:1755]: Setting this to `true` is only required when using instance IP address auto-detection, right? (ie. if an IPv6 address is explicitly given to `-query-frontend.instance-addr`, then it doesn't matter what value this option has)

If so, could you please clarify the description and changelog entry to be clear that this is only relevant when an explicit address is not given?

Comment (Luke-Smartnews) [reply cmd/mimir/help-all.txt.tmpl:1755]: sure, although I just used the same description with other components like [compactor](https://grafana.com/docs/mimir/latest/references/configuration-parameters/#compactor).

Comment (Luke-Smartnews) [reply cmd/mimir/help-all.txt.tmpl]: Done

Comment (charleskorn) [reply cmd/mimir/help-all.txt.tmpl]: Ah, I didn't realise we used that description elsewhere sorry - in that case, it's probably better to be consistent.

Comment (Luke-Smartnews) [reply cmd/mimir/help-all.txt.tmpl]: Do you want to me to revert the change?

Comment (charleskorn) [reply cmd/mimir/help-all.txt.tmpl]: Sorry, yes please.

Comment (Luke-Smartnews) [reply cmd/mimir/help-all.txt.tmpl]: done

**Review by charleskorn — APPROVED** [review]

**Pull request merged**